File: internal/signals/signals.go
// Package signals derives the signals the model reads from a card's history.
//
// Every window and every gap is measured against event time -- the timestamp
// carried on the authorization -- never against the clock at which scoring
// happens, so a replayed authorization produces the values it produced live.
//
// The training side states the same definitions in training/signals.py. Nothing
// makes the two agree except the train/serve fixture, which is why that fixture
// is a build artifact rather than a belief; see scoring's fixture test.
package signals

import (
	"math"
	"sort"
)

// Set is the signals the model reads, in the order the model reads them. This
// is the one place the set is stated on the serving side; adding a signal is an
// edit here, in Derive, and in training/signals.py.
var Set = []string{"amount", "gap_seconds", "velocity_1h", "velocity_24h"}

// Velocity counter windows, in seconds. A payment sitting exactly Window
// seconds before the authorization is counted: the lower bound is inclusive.
var velocityWindows = []struct {
	Name   string
	Window int64
}{
	{"velocity_1h", 3600},
	{"velocity_24h", 86400},
}

// A Payment is one entry in a card's history, identified well enough to be
// ordered against the others.
type Payment struct {
	TransactionID string
	EventTime     int64 // unix seconds
}

// A Timeline is one card's payments in key order: event time, then transaction
// id. Whole-second timestamp resolution puts genuinely distinct payments at the
// same event time, so the transaction id is what leaves "the previous payment"
// defined (ADR-0006).
type Timeline []Payment

func (t Timeline) Len() int      { return len(t) }
func (t Timeline) Swap(i, j int) { t[i], t[j] = t[j], t[i] }
func (t Timeline) Less(i, j int) bool {
	if t[i].EventTime != t[j].EventTime {
		return t[i].EventTime < t[j].EventTime
	}
	return t[i].TransactionID < t[j].TransactionID
}

// Derive assembles every signal in Set for one authorization.
//
// Only payments strictly before the authorization in key order are read, so a
// signal never contains information from its own outcome, and history is that
// one card's own, so interleaved traffic across cards cannot contaminate a
// counter. A missing gap -- a card's first payment -- is NaN rather than a
// number the model would read as a very short gap.
func Derive(history Timeline, auth Payment, amountUSD float64) map[string]float64 {
	predecessors := history[:sort.Search(len(history), func(i int) bool {
		return !before(history[i], auth)
	})]

	derived := map[string]float64{
		"amount":      amountUSD,
		"gap_seconds": math.NaN(),
	}
	if n := len(predecessors); n > 0 {
		derived["gap_seconds"] = float64(auth.EventTime - predecessors[n-1].EventTime)
	}
	for _, counter := range velocityWindows {
		derived[counter.Name] = float64(len(predecessors) - firstInWindow(predecessors, auth.EventTime, counter.Window))
	}
	return derived
}

// firstInWindow is the index of the earliest payment within `window` seconds
// before `at`, inclusive of the boundary itself.
func firstInWindow(predecessors Timeline, at, window int64) int {
	return sort.Search(len(predecessors), func(i int) bool {
		return predecessors[i].EventTime >= at-window
	})
}

func before(a, b Payment) bool {
	if a.EventTime != b.EventTime {
		return a.EventTime < b.EventTime
	}
	return a.TransactionID < b.TransactionID
}

// Vector lays derived signals out in Set order, which is the order the model
// was trained to read them in.
func Vector(derived map[string]float64) []float64 {
	vector := make([]float64, len(Set))
	for i, name := range Set {
		vector[i] = derived[name]
	}
	return vector
}

File: training/signals.py
"""Signal derivation, training side.

The signals the model reads are stated once, here, in `SIGNAL_SET`. The serving
side states them once too, in `internal/signals`, and the train/serve fixture is
what forces the two statements to agree -- see `testdata/train_serve_fixture.json`.

Every window and every gap is measured against event time (the timestamp carried
on the authorization), never against the clock. Payments are ordered by event
time then transaction id, because whole-second timestamp resolution puts
genuinely distinct payments at the same event time; see ADR-0006.
"""

import numpy as np
import pandas as pd

# The signals the model reads, in the order the model reads them. One place.
SIGNAL_SET = ["amount", "gap_seconds", "velocity_1h", "velocity_24h"]

# Velocity counter windows, in seconds. A payment sitting exactly `window`
# seconds before the authorization is counted: the lower bound is inclusive.
VELOCITY_WINDOWS = {"velocity_1h": 3600, "velocity_24h": 86400}

_COLUMNS = ["cc_num", "trans_num", "unix_time", "amt", "is_fraud"]


def load_payments(paths):
    """Read the checked-in CSVs into one frame, ordered by card then event time.

    -> DataFrame[card, transaction_id, event_time, amount, is_fraud]
    """
    frames = []
    for path in paths:
        frame = pd.read_csv(path, usecols=_COLUMNS, dtype={"cc_num": str})
        frames.append(frame)
    payments = pd.concat(frames, ignore_index=True)
    payments = payments.rename(columns={
        "cc_num": "card",
        "trans_num": "transaction_id",
        "unix_time": "event_time",
        "amt": "amount",
    })
    payments["event_time"] = payments["event_time"].astype(np.int64)
    payments["amount"] = payments["amount"].astype(np.float64)
    payments["is_fraud"] = payments["is_fraud"].astype(np.int8)
    return _in_key_order(payments)


def _in_key_order(payments):
    """Card, then event time, then transaction id. The ordering, everywhere."""
    return payments.sort_values(
        ["card", "event_time", "transaction_id"], kind="mergesort"
    ).reset_index(drop=True)


def derive(payments):
    """Add every signal in SIGNAL_SET to a frame already in key order.

    Each card's signals come only from that card's own history, and only from
    payments strictly before the authorization being scored, so a signal never
    contains information from its own outcome.
    """
    payments = payments.copy()
    payments["amount"] = payments["amount"].astype(np.float64)

    gap = np.full(len(payments), np.nan)
    velocity = {name: np.zeros(len(payments), dtype=np.float64)
                for name in VELOCITY_WINDOWS}

    for start, stop in card_spans(payments["card"].to_numpy()):
        times = payments["event_time"].to_numpy()[start:stop]

        # The gap to the previous payment on this card. The first payment on a
        # card has no previous payment, so its gap stays missing rather than
        # becoming a number the model would read as small.
        gap[start + 1:stop] = np.diff(times).astype(np.float64)

        # Velocity counters. Predecessors are the earlier rows in key order, so
        # counting them is index arithmetic: everything from the first row
        # inside the window up to (but not including) this one.
        index = np.arange(stop - start)
        for name, window in VELOCITY_WINDOWS.items():
            first_in_window = np.searchsorted(times, times - window, side="left")
            velocity[name][start:stop] = (index - first_in_window).astype(np.float64)

    payments["gap_seconds"] = gap
    for name, counts in velocity.items():
        payments[name] = counts
    return payments


def card_spans(cards):
    """-> [(start, stop), ...] row spans, one per card, over a card-ordered array."""
    if len(cards) == 0:
        return []
    boundaries = np.flatnonzero(cards[1:] != cards[:-1]) + 1
    edges = np.concatenate(([0], boundaries, [len(cards)]))
    return list(zip(edges[:-1], edges[1:]))


def signal_matrix(payments):
    """-> the SIGNAL_SET columns, in order, as the model expects them."""
    return payments[SIGNAL_SET]
